js-api-doc/importer.d.ts

```diff
@@ -3,32 +3,293 @@ import {URL} from 'url';
+   *
+   * Calling [[canonicalize]] multiple times with the same URL must return the
+   * same result. Calling [[canonicalize]] with a URL returned by a previous
+   * call to [[canonicalize]] must return that URL.
+   *
+   * Relative loads in stylesheets loaded from an importer are handled by
+   * resolving the loaded URL relative to the canonical URL of the stylesheet
+   * that contains it, and passing that URL back to the importer's
+   * [[canonicalize]] method. For example, suppose the "Resolving a Load"
+   * example {@link Importer | above} returned a stylesheet that contained
+   * `@use "mixins"`:
+   *
+   * - The compiler resolves the URL `mixins` relative to the current
+   *   stylesheet's canonical URL `db:foo/bar/baz/_index.scss` to get
+   *   `db:foo/bar/baz/mixins`.
+   * - It calls [[canonicalize]] with `"db:foo/bar/baz/mixins"`.
+   * - [[canonicalize]] returns `new URL("db:foo/bar/baz/_mixins.scss")`.
+   *
+   * Because of this, [[canonicalize]] must return a meaningful result when
+   * called with a URL relative to one returned by an earlier call to
+   * [[canonicalize]].
+   *
+   * @param url - The loaded URL. Since this might be relative, it's represented
+   * as a string rather than a [[URL]] object.
+   *
+   * @param options.fromImport - Whether this is being invoked because of a Sass
+   * `@import` rule, as opposed to a `@use` or `@forward` rule.
+   *
+   * This should *only* be used for determining whether or not to load
+   * [import-only files](https://sass-lang.com/documentation/at-rules/import#import-only-files).
+   *
+   * @returns An absolute URL if this importer recognizes the `url`, or `null`
+   * if it doesn't. If this returns `null`, other importers or {@link
+   * Options.loadPaths | load paths} may handle the load.
+   *
+   * This may also return a `Promise`, but if it does the importer may only be
+   * passed to [[compileAsync]] and [[compileStringAsync]], not [[compile]] or
+   * [[compileString]].
+   *
+   * @throws any - If this importer recognizes the URL but determines that
+   * it's invalid, it may throw an error. This error will be wrapped in an
+   * [[Exception]] associated with the offending load rule.
+   */
   canonicalize(
     url: string,
     options: {fromImport: boolean}
   ): PromiseOr<URL | null, sync>;
 
+  /**
+   * Loads the Sass text for the given `canonicalUrl`, or returns `null` if this
+   * importer can't find the stylesheet it refers to.
+   *
+   * If the importer finds a stylesheet at [url] but it fails to load for some
+   * reason, or if [url] is uniquely associated with this importer but doesn't
+   * refer to a real stylesheet, the importer may throw an exception that will
+   * be wrapped by Sass. If the exception object has a `message` property, it
+   * will be used as the wrapped exception's message; otherwise, the exception
+   * object's `toString()` will be used. This means it's safe for importers to
+   * throw plain strings.
```

Comment:
Ah... this answers my earlier question about throwing `any`. Might be worth duplicating this in the other jsdocs.

Reply:
Good thought, I've updated all three to use `@throws` annotations with more detail.